packages/box/src/Box.tsx

```diff
@@ -1,68 +1,69 @@
+type InferComponentType<C, H> = H extends string
+  ? 'a' | React.ComponentType
+  : C;
 
-type AnchorProps<T> = HTMLElementProps<'a'> &
-  T & {
-    href: string;
-  };
-
-type CustomElementProps<T> = T & {
-  component: React.ElementType<any>;
-  [key: string]: any;
-};
-
-export type BoxProps<T> = DivProps<T> | AnchorProps<T> | CustomElementProps<T>;
-
-function isCustomElement<T>(
-  props: BoxProps<T>,
-): props is CustomElementProps<T> {
-  return (props as any).component != null;
+interface BoxComponentProps<
+  C extends React.ElementType = React.ElementType,
+  H extends string | undefined = undefined
+> {
+  href?: H;
+  as?: InferComponentType<C, H>;
 }
 
-function isAnchorElement<T>(props: BoxProps<T>): props is AnchorProps<T> {
-  return (props as any).href != null;
-}
-
-/**
- * # Box
- *
- * Box component
- *
- * ```
-<Box href="https://mongodb.design">Anchors Away!</Box>
-```
- * @param props.children Content to be rendered in an HTML element, or provided as a prop to the rendered component.
- * @param props.href When provided, `<Box />` will render an anchor tag with this `href` value.
- * @param props.component The component or HTML tag to be rendered by the `<Box />` component. **Note**: This will supersede the behavior of any other props.
- */
+type BoxProps<
+  C extends React.ElementType,
+  H extends string | undefined
+> = BoxComponentProps<C, H> &
+  Omit<
+    React.ComponentPropsWithRef<C extends React.ComponentType ? any : C>,
+    keyof BoxComponentProps
+  >;
 
-const Box = React.forwardRef(
-  <T extends React.ReactNode>(props: BoxProps<T>, ref: React.Ref<any>) => {
-    const rest = omit(props as any, ['component']);
-    let Box: React.ElementType<any> = 'div';
+type BoxType = <
+  C extends React.ElementType = 'div',
+  H extends string | undefined = undefined
+>(
+  props: BoxProps<InferComponentType<C, H>, H>,
+) => JSX.Element;
 
-    if (isCustomElement<T>(props)) {
-      Box = props.component;
-    } else if (isAnchorElement<T>(props)) {
-      Box = 'a';
-    }
+// eslint-disable-next-line
+export const Box = React.forwardRef(
+  <
+    C extends React.ElementType = 'div',
+    H extends string | undefined = undefined
+  >(
+    { as, href, ...rest }: BoxProps<C, H>,
+    ref: React.Ref<any>,
+  ) => {
+    const Component = as ? as : href ? 'a' : 'div';
 
-    return <Box {...rest} ref={ref} />;
+    return <Component href={href} ref={ref} {...rest} />;
   },
-);
+) as BoxType;
 
+// @ts-ignore Property 'displayName' does not exist on type 'BoxType'.
 Box.displayName = 'Box';
 
-// @ts-ignore: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/37660
+// @ts-ignore Property 'propTypes' does not exist on type 'BoxType'
 Box.propTypes = {
-  children: PropTypes.node,
+  as: PropTypes.elementType,
   href: PropTypes.string,
-  // @ts-ignore
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
 };
 
 export default Box;
+
+export type OverrideComponentProps<
```

Comment:
Locally, I was able to eliminate this type entirely by modifying `BoxProps` like so:
```ts
export type BoxProps<
  C extends React.ElementType,
  H extends string | undefined,
  P extends Record<string, any> = {}
> = P & BoxComponentProps<C, H> &
  Omit<
    React.ComponentPropsWithRef<C extends React.ComponentType ? any : C>,
    keyof BoxComponentProps
  >;
```

And then replacing all instances of `OverrideComponentProps` with it.

Reply:
👍  Done